File: code401challenges/src/main/java/tree/BinarySearchTree.java
package tree;

import java.rmi.AlreadyBoundException;

public class BinarySearchTree {
    //instance variables
    Node root;

    //Constructor function
    BinarySearchTree () {
        this.root = null;
    }

    BinarySearchTree (Node node) {
        this.root = node;
    }

    //instance methods
    public void add (int value) {
        this.root = add(value, this.root);
    }

    public Node add (int value, Node root) {
        if (root == null) {
            return new Node(value);
        } else {

            if (value < root.value) {
               root.left = add(value, root.left);
            } else if (value > root.value) {
                root.right = add(value, root.right);
            } else {
                throw new IllegalArgumentException("Value already exists in the tree");
            }
        }
        return root;
    }

    public boolean contains (int value) {
        return containsRecursive(value, this.root);
    }

    public boolean containsRecursive (int value, Node node) {
        if (node == null) {
            return false;
        } else {
            if (value == node.value) {
                return true;
            } else if (value < node.value) {
                return containsRecursive(value, node.left);
            } else {
                return containsRecursive(value, node.right);
            }
        }
    }
}

File: js401challenges/Trees/BinarySearchTree.js
'use strict';

const BinaryTree = require('./BinaryTree.js');

class BinarySearchTree extends BinaryTree {

  constructor (root = null) {
    super(root);
  }

  add (value) {
    return this.root = this.addHelper(value, this.root);
  }

  addHelper (value, root) {
    if (root === null) {
      return new Node(value);
    } else {

      if (value < root.value) {
        root.left = this.addHelper(value, root.left);
      } else if (value > root.value) {
        root.right = this.addHelper(value, root.right);
      } else {
        throw new Error('Value already exists in the code401challenges.tree');
      }
    }
    return root;
  }

  contains (value) {
    return this.containsHelper(value, this.root);
  }

  containsHelper (value, node) {
    if (node === null) {
      return false;
    } else {
      if (value === node.value) {
        return true;
      } else if (value < node.value) {
        return this.containsHelper(value, node.left);
      } else {
        return this.containsHelper(value, node.right);
      }
    }
  }

}

module.exports = BinarySearchTree;
